proposals/0069_prometheus-exporters-otel-embedding.md

```diff
@@ -0,0 +1,286 @@
+}
+```
+
+### Prometheus Registry to pmetric Conversion
+
+The adapter will include a scraper component that:
+
+1. Calls `registry.Gather()` to collect metrics from the Prometheus Registry
+2. Converts Prometheus metric families to OpenTelemetry's pmetric format
+
+This conversion logic can leverage or adapt existing conversion code from the OpenTelemetry Prometheus receiver.
+
+### OpenTelemetry Collector Receiver Implementation
+
+The `exporter-toolkit` will provide a complete implementation of OpenTelemetry's receiver interfaces:
+
+1. **component.Factory** - for component type and default configuration
+2. **component.Component** - for lifecycle management
+3. **receiver.Factory** - for creating receiver instances
+4. **receiver.Metrics** - for producing pmetric data
+
+This implementation will:
+- Start the Prometheus exporter and obtain its Registry
+- Run a periodic scrape loop based on the configured interval
+- Convert scraped metrics to pmetric format
+- Push metrics to the OpenTelemetry pipeline consumer
+
+### Using with OpenTelemetry Collector Builder
+
+Once a Prometheus exporter implements the new interfaces, it can be included in custom OpenTelemetry Collector distributions via OCB:
+
+```yaml
+# ocb-config.yaml
+receivers:
+  - gomod: github.com/prometheus/node_exporter v1.x.x
+```
+
+The OCB will recognize the exporter as a valid receiver and include it in the built collector binary.
+
+### Example Configuration
+
+In the OpenTelemetry Collector configuration:
+
+```yaml
+receivers:
+  node_exporter:
+    scrape_interval: 30s
+    exporter_config:
+      # Node exporter specific configuration
+      collectors:
+        - cpu
+        - diskstats
+        - filesystem
+
+exporters:
+  otlp:
+    endpoint: otelcol:4317
+
+service:
+  pipelines:
+    metrics:
+      receivers: [node_exporter]
+      exporters: [otlp]
+```
+
+### Implementation Steps
+
+1. **Extend exporter-toolkit** with the new interfaces and OpenTelemetry Collector receiver implementation
+2. **Implement the Prometheus to pmetric converter** (potentially adapting existing code)
+3. **Update one reference exporter** (e.g., node_exporter) to implement the new interfaces as a proof of concept
+4. **Validate** with OpenTelemetry Collector Builder
+5. **Document** the integration pattern for other exporters
+6. **Gradually adopt** across other Prometheus exporters based on community interest
+
+### Migration and Compatibility
+
+* **No breaking changes** to existing Prometheus exporters that don't adopt the interfaces
+* **Opt-in adoption** - exporters can choose when/if to implement embedding support
+* **Backward compatibility** - embedded exporters still work as standalone exporters
+
+
+### Known Problems
+
+1. **Dependency conflicts**: Prometheus exporters and OpenTelemetry collector-contrib use different dependency versions. Building a distribution with both may require dependency alignment or replace directives.
```

Comment:
We need to be careful to not accidentally link in half of the OTel ecosystem into a normal exporter build. In principle, I assume that the exporter build should be mostly unaffected, but we know how things go if you do some innocuous import of some client library…

Reply:
It's inevitable to introduce otel dependencies while building the adapter, but I see three ways forward:

* A Prometheus exporter does not implement the interfaces and continues not to be embeddable
* A Prometheus exporter implements the interfaces in its main module and inherits the otel dependencies
* A Prometheus exporter creates a separate module that will be embeddable. The original module remains intact